File: ProAssetin.API/Services/ProjectService.cs
using Microsoft.EntityFrameworkCore;
using ProAssetin.API.Data;
using ProAssetin.API.Models;
using ProAssetin.API.Models.DTOs;

namespace ProAssetin.API.Services
{
    public class ProjectService : IProjectService
    {
        private readonly ITenantDbContextFactory _dbContextFactory;
        private readonly ApplicationDbContext _masterContext;

        public ProjectService(ITenantDbContextFactory dbContextFactory, ApplicationDbContext masterContext)
        {
            _dbContextFactory = dbContextFactory;
            _masterContext = masterContext;
        }

        public async Task<(IEnumerable<ProjectDto> Items, int TotalCount)> GetProjectsAsync(ProjectQueryDto query, string tenantId)
        {
            using var context = _dbContextFactory.CreateDbContext(tenantId);
            var tid = tenantId?.ToLowerInvariant();

            var q = context.Projects.Where(p => p.TenantId.ToLower() == tid).AsQueryable();

            if (!string.IsNullOrWhiteSpace(query.SearchTerm))
            {
                var t = query.SearchTerm;
                q = q.Where(p =>
                    p.ProjectReference.Contains(t) ||
                    p.Name.Contains(t) ||
                    (p.Description != null && p.Description.Contains(t)) ||
                    (p.ProjectManagerName != null && p.ProjectManagerName.Contains(t)) ||
                    (p.DepartmentOrClient != null && p.DepartmentOrClient.Contains(t)));
            }

            if (!string.IsNullOrWhiteSpace(query.Status))
                q = q.Where(p => p.Status == query.Status);

            if (!string.IsNullOrWhiteSpace(query.Priority))
                q = q.Where(p => p.Priority == query.Priority);

            if (query.StartDateFrom.HasValue)
                q = q.Where(p => p.StartDate >= query.StartDateFrom.Value);

            if (query.StartDateTo.HasValue)
                q = q.Where(p => p.StartDate <= query.StartDateTo.Value);

            var total = await q.CountAsync();

            q = query.SortBy?.ToLower() switch
            {
                "name" => query.SortDescending ? q.OrderByDescending(p => p.Name) : q.OrderBy(p => p.Name),
                "projectreference" => query.SortDescending ? q.OrderByDescending(p => p.ProjectReference) : q.OrderBy(p => p.ProjectReference),
                "status" => query.SortDescending ? q.OrderByDescending(p => p.Status) : q.OrderBy(p => p.Status),
                "priority" => query.SortDescending ? q.OrderByDescending(p => p.Priority) : q.OrderBy(p => p.Priority),
                "startdate" => query.SortDescending ? q.OrderByDescending(p => p.StartDate) : q.OrderBy(p => p.StartDate),
                "enddate" => query.SortDescending ? q.OrderByDescending(p => p.EndDate) : q.OrderBy(p => p.EndDate),
                _ => query.SortDescending ? q.OrderByDescending(p => p.CreatedAt) : q.OrderByDescending(p => p.CreatedAt)
            };

            var rows = await q
                .Skip((query.PageNumber - 1) * query.PageSize)
                .Take(query.PageSize)
                .ToListAsync();

            var userIds = rows.Where(p => !string.IsNullOrEmpty(p.CreatedByUserId)).Select(p => p.CreatedByUserId!).Distinct().ToList();
            var names = await ResolveUserNamesAsync(userIds);
            return (rows.Select(p => MapToDto(p, names)), total);
        }

        public async Task<ProjectDto?> GetProjectByIdAsync(int id, string tenantId)
        {
            using var context = _dbContextFactory.CreateDbContext(tenantId);
            var tid = tenantId?.ToLower();

            var row = await context.Projects
                .FirstOrDefaultAsync(p => p.Id == id && p.TenantId.ToLower() == tid);

            if (row == null) return null;

            var names = await ResolveUserNamesAsync(
                string.IsNullOrEmpty(row.CreatedByUserId) ? new List<string>() : new List<string> { row.CreatedByUserId! });

            return MapToDto(row, names);
        }

        public async Task<ProjectDto> CreateProjectAsync(CreateProjectDto dto, string tenantId, string userId)
        {
            using var context = _dbContextFactory.CreateDbContext(tenantId);
            var tid = tenantId?.ToLowerInvariant()!;

            var row = new Project
            {
                ProjectReference = dto.ProjectReference,
                Name = dto.Name,
                Description = dto.Description,
                Status = string.IsNullOrWhiteSpace(dto.Status) ? "Planning" : dto.Status,
                Priority = string.IsNullOrWhiteSpace(dto.Priority) ? "Medium" : dto.Priority,
                StartDate = dto.StartDate,
                EndDate = dto.EndDate,
                ProjectManagerName = dto.ProjectManagerName,
                DepartmentOrClient = dto.DepartmentOrClient,
                Notes = dto.Notes,
                TenantId = tid,
                CreatedByUserId = userId,
                CreatedAt = DateTime.UtcNow
            };

            context.Projects.Add(row);
            await context.SaveChangesAsync();

            var names = await ResolveUserNamesAsync(new List<string> { userId });
            return MapToDto(row, names);
        }

        public async Task<ProjectDto?> UpdateProjectAsync(int id, UpdateProjectDto dto, string tenantId)
        {
            using var context = _dbContextFactory.CreateDbContext(tenantId);
            var tid = tenantId?.ToLower();

            var row = await context.Projects
                .FirstOrDefaultAsync(p => p.Id == id && p.TenantId.ToLower() == tid);

            if (row == null) return null;

            if (!string.IsNullOrWhiteSpace(dto.ProjectReference))
                row.ProjectReference = dto.ProjectReference;

            if (!string.IsNullOrWhiteSpace(dto.Name))
                row.Name = dto.Name;

            if (dto.Description != null)
                row.Description = dto.Description;

            if (!string.IsNullOrWhiteSpace(dto.Status))
                row.Status = dto.Status;

            if (!string.IsNullOrWhiteSpace(dto.Priority))
                row.Priority = dto.Priority;

            if (dto.StartDate.HasValue)
                row.StartDate = dto.StartDate;

            if (dto.EndDate.HasValue)
                row.EndDate = dto.EndDate;

            if (dto.ProjectManagerName != null)
                row.ProjectManagerName = dto.ProjectManagerName;

            if (dto.DepartmentOrClient != null)
                row.DepartmentOrClient = dto.DepartmentOrClient;

            if (dto.Notes != null)
                row.Notes = dto.Notes;

            row.UpdatedAt = DateTime.UtcNow;
            await context.SaveChangesAsync();

            var names = await ResolveUserNamesAsync(
                string.IsNullOrEmpty(row.CreatedByUserId) ? new List<string>() : new List<string> { row.CreatedByUserId! });

            return MapToDto(row, names);
        }

        public async Task<bool> DeleteProjectAsync(int id, string tenantId)
        {
            using var context = _dbContextFactory.CreateDbContext(tenantId);
            var tid = tenantId?.ToLower();

            var row = await context.Projects
                .FirstOrDefaultAsync(p => p.Id == id && p.TenantId.ToLower() == tid);

            if (row == null) return false;

            context.Projects.Remove(row);
            await context.SaveChangesAsync();
            return true;
        }

        public Task<IEnumerable<string>> GetStatusesAsync()
        {
            return Task.FromResult<IEnumerable<string>>(new[]
            {
                "Planning", "Active", "On Hold", "Completed", "Cancelled"
            });
        }

        public Task<IEnumerable<string>> GetPrioritiesAsync()
        {
            return Task.FromResult<IEnumerable<string>>(new[] { "Low", "Medium", "High" });
        }

        private async Task<Dictionary<string, string>> ResolveUserNamesAsync(List<string> userIds)
        {
            if (!userIds.Any()) return new Dictionary<string, string>();

            var users = await _masterContext.Users
                .Where(u => userIds.Contains(u.Id))
                .Select(u => new { u.Id, u.FirstName, u.LastName })
                .ToListAsync();

            return users.ToDictionary(u => u.Id, u => $"{u.FirstName} {u.LastName}");
        }

        private static ProjectDto MapToDto(Project p, Dictionary<string, string> userNames)
        {
            string? createdName = null;
            if (!string.IsNullOrEmpty(p.CreatedByUserId))
                userNames.TryGetValue(p.CreatedByUserId, out createdName);

            return new ProjectDto
            {
                Id = p.Id,
                ProjectReference = p.ProjectReference,
                Name = p.Name,
                Description = p.Description,
                Status = p.Status,
                Priority = p.Priority,
                StartDate = p.StartDate,
                EndDate = p.EndDate,
                ProjectManagerName = p.ProjectManagerName,
                DepartmentOrClient = p.DepartmentOrClient,
                Notes = p.Notes,
                CreatedByUserId = p.CreatedByUserId,
                CreatedByUserName = createdName,
                CreatedAt = p.CreatedAt,
                UpdatedAt = p.UpdatedAt
            };
        }
    }
}

File: ProAssetin.Web/src/app/core/services/project.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface Project {
  id: number;
  projectReference: string;
  name: string;
  description?: string | null;
  status: string;
  priority: string;
  startDate?: string | null;
  endDate?: string | null;
  projectManagerName?: string | null;
  departmentOrClient?: string | null;
  notes?: string | null;
  createdByUserName?: string | null;
  createdAt: string;
  updatedAt?: string | null;
}

export interface CreateProject {
  projectReference: string;
  name: string;
  description?: string;
  status?: string;
  priority?: string;
  startDate?: string | null;
  endDate?: string | null;
  projectManagerName?: string;
  departmentOrClient?: string;
  notes?: string;
}

export interface UpdateProject {
  projectReference?: string;
  name?: string;
  description?: string;
  status?: string;
  priority?: string;
  startDate?: string | null;
  endDate?: string | null;
  projectManagerName?: string;
  departmentOrClient?: string;
  notes?: string;
}

export interface ProjectQuery {
  pageNumber?: number;
  pageSize?: number;
  searchTerm?: string;
  status?: string;
  priority?: string;
  startDateFrom?: string;
  startDateTo?: string;
  sortBy?: string;
  sortDescending?: boolean;
}

export interface ProjectListResponse {
  data: Project[];
  totalCount: number;
  pageNumber: number;
  pageSize: number;
}

@Injectable({ providedIn: 'root' })
export class ProjectService {
  private base = `${environment.apiUrl}/projects`;

  constructor(private http: HttpClient) {}

  getProjects(q: ProjectQuery = {}): Observable<ProjectListResponse> {
    let p = new HttpParams();
    if (q.pageNumber) p = p.set('pageNumber', String(q.pageNumber));
    if (q.pageSize) p = p.set('pageSize', String(q.pageSize));
    if (q.searchTerm) p = p.set('searchTerm', q.searchTerm);
    if (q.status) p = p.set('status', q.status);
    if (q.priority) p = p.set('priority', q.priority);
    if (q.startDateFrom) p = p.set('startDateFrom', q.startDateFrom);
    if (q.startDateTo) p = p.set('startDateTo', q.startDateTo);
    if (q.sortBy) p = p.set('sortBy', q.sortBy);
    if (q.sortDescending !== undefined) p = p.set('sortDescending', String(q.sortDescending));
    return this.http.get<ProjectListResponse>(this.base, { params: p });
  }

  getProject(id: number): Observable<Project> {
    return this.http.get<Project>(`${this.base}/${id}`);
  }

  createProject(body: CreateProject): Observable<Project> {
    return this.http.post<Project>(this.base, body);
  }

  updateProject(id: number, body: UpdateProject): Observable<Project> {
    return this.http.put<Project>(`${this.base}/${id}`, body);
  }

  deleteProject(id: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/${id}`);
  }

  getStatuses(): Observable<string[]> {
    return this.http.get<string[]>(`${this.base}/statuses`);
  }

  getPriorities(): Observable<string[]> {
    return this.http.get<string[]>(`${this.base}/priorities`);
  }
}
